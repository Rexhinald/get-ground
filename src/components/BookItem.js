import React, {useState} from "react";
import {Collapse, List, ListItem, ListItemButton, ListItemText} from "@mui/material";
import {ExpandLess, ExpandMore} from "@mui/icons-material";

const BookItem = ({book}) => {
  const [showDetails, setShowDetails] = useState(false);

  const toggle = () => {
    setShowDetails(prevState => !prevState);
  };

  const authors = () => (
    book.book_author.map((author, index) => {
      return `${author}${index !== book.book_author.length - 1 ? ',' : ''} `;
    })
  );

  return (
    <>
      <ListItemButton data-testid="book-item" onClick={toggle}>
        <ListItemText primary={book.book_title} />
        {showDetails ? <ExpandLess /> : <ExpandMore />}
      </ListItemButton>
      <Collapse data-testid="book-details" in={showDetails} unmountOnExit>
        <List component="div" sx={{pl: 3}}>
          <ListItem>
            <ListItemText>Author: {authors()}</ListItemText>
          </ListItem>
          <ListItem>
            <ListItemText>Pages: {book.book_pages}</ListItemText>
          </ListItem>
          <ListItem>
            <ListItemText>Publication place: {book.book_publication_city}, {book.book_publication_country}</ListItemText>
          </ListItem>
          <ListItem>
            <ListItemText>Publication year: {book.book_publication_year}</ListItemText>
          </ListItem>
        </List>
      </Collapse>
    </>
  );
};

export default BookItem;
