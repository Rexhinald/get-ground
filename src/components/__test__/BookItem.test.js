import React from "react";
import ReactDOM from "react-dom";
import BookItem from "../BookItem";
import {cleanup, fireEvent, render} from "@testing-library/react";
import renderer from "react-test-renderer";

afterEach(cleanup);

const book = {
  "id": 2149,
  "book_author": [
    "Δωρόθεος μητροπολίτης Μονεμβασίας"
  ],
  "book_title": "Βιβλίον Ιστορικόν",
  "book_publication_year": 1654,
  "book_publication_country": "Ιταλία",
  "book_publication_city": "Βενετία",
  "book_pages": 0
};

it('renders without crashing', () => {
  const div = document.createElement("div");
  ReactDOM.render(<BookItem book={book} />, div);
});

it('renders book details', () => {
  const {getByTestId} = render(<BookItem book={book} />);
  fireEvent(getByTestId('book-item'), new MouseEvent("click", {
    bubbles: true,
    cancelable: true,
  }));
  expect(getByTestId('book-details')).toBeTruthy();
});

it('matches snapshot', () => {
  const tree = renderer.create(<BookItem book={book} />).toJSON();
  expect(tree).toMatchSnapshot();
});
