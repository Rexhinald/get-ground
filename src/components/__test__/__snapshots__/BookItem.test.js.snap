// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
  data-testid="book-item"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="MuiListItemText-root css-tlelie-MuiListItemText-root"
  >
    <span
      className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
    >
      Βιβλίον Ιστορικόν
    </span>
  </div>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="ExpandMoreIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
    />
  </svg>
</div>
`;
